# Data Engineering JDA
#### Created By: Prateek Rawat
#### Date: 05/18/2019

## Data Description
    1. Data is about NYC Taxi & Limousine Commission YELLOW TAXI trip record.
    2. Year wise data is presnt at https://www1.nyc.gov/site/tlc/about/tlc-trip-record-data.page
## Task To Be Performed
    1. Create a simple data pipeline to fetch data from NYC TLC website for Yellow Taxi.
    2. Write a Python program that calculates the average/mean trip length of all Yellow Taxi for a month.
    3. Extend this data piple that can ingest new data and calculates the 45 days rolling average trip length.
## Task Performed
    1. Ingested data from website for January 2018 and February 2018.
    2. Datatype Conversion.
    3. Creating new field to perform analysis efficeintly.
    4. Creating Functions for cleaning of redudant value and calculating mean.
## Output
    As it is unclear that whether average trip length is for trip distance or for trip time.I have perfomed mean calculation analysis for both the variables.
    

In [1]:
#Importing Libraries
import pandas as pd
import io
import requests
import warnings
warnings.filterwarnings('ignore')

#### Function taxi is created to ingest data from https://www1.nyc.gov/site/tlc/about/tlc-trip-record-data.page 
    1. Library 'requests' is used to send http request to call url with 'get' method.
    2. The ingested .csv data is decoded as 'utf-8'

In [2]:
# Function to call destination(URL) to copy .csv into Pandas DataFrame
df= pd.DataFrame()
def taxi(year,month,df=df):
    base_url = 'https://s3.amazonaws.com/nyc-tlc/trip+data/yellow_tripdata_'
    l=base_url+year+'-'+month +'.csv'  
    s=requests.get(l).content
    if df.empty:
        df = pd.read_csv(io.StringIO(s.decode('utf-8')))
    else:
        df = df.append(pd.read_csv(io.StringIO(s.decode('utf-8'))), ignore_index = True)
    return df

In [3]:
# Calling Function to download data for January 2018
df30Days = taxi('2018','01')

In [4]:
#Checking Data
df30Days.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2018-01-01 00:21:05   2018-01-01 00:24:23                1   
1         1  2018-01-01 00:44:55   2018-01-01 01:03:05                1   
2         1  2018-01-01 00:08:26   2018-01-01 00:14:21                2   
3         1  2018-01-01 00:20:22   2018-01-01 00:52:51                1   
4         1  2018-01-01 00:09:18   2018-01-01 00:27:06                2   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            0.5           1                  N            41            24   
1            2.7           1                  N           239           140   
2            0.8           1                  N           262           141   
3           10.2           1                  N           140           257   
4            2.5           1                  N           246           239   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          4.5    0.5      0.5        0.00           0.0   
1             2         14.0    0.5      0.5        0.00           0.0   
2             1          6.0    0.5      0.5        1.00           0.0   
3             2         33.5    0.5      0.5        0.00           0.0   
4             1         12.5    0.5      0.5        2.75           0.0   

   improvement_surcharge  total_amount  
0                    0.3          5.80  
1                    0.3         15.30  
2                    0.3          8.30  
3                    0.3         34.80  
4                    0.3         16.55

In [5]:
# Calling taxi function to dowload data for 2 months
df60Days = taxi('2018', '02', df30Days)

In [6]:
# Checking Data From Tail
df60Days.tail()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  \
17251945         2  2018-02-28 23:21:56   2018-02-28 23:47:43   
17251946         2  2018-02-28 23:54:07   2018-03-01 00:24:56   
17251947         2  2018-02-28 23:17:42   2018-02-28 23:53:11   
17251948         2  2018-03-01 13:39:17   2018-03-01 13:49:08   
17251949         2  2018-02-28 23:03:18   2018-02-28 23:03:24   

          passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
17251945                6           9.61           1                  N   
17251946                6           4.20           1                  N   
17251947                1           0.00           1                  N   
17251948                2           0.00           1                  N   
17251949                1           0.00           1                  N   

          PULocationID  DOLocationID  payment_type  fare_amount  extra  \
17251945           138           163             2         28.5    0.5   
17251946           230           230             1         21.5    0.5   
17251947           193           193             2          0.0    0.0   
17251948           264           264             2          2.4    0.0   
17251949           264           193             1          0.0    0.0   

          mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
17251945      0.5         0.0          5.54                    0.3   
17251946      0.5         0.0          0.00                    0.3   
17251947      0.0         0.0          0.00                    0.0   
17251948      0.5         0.0          0.00                    0.3   
17251949      0.0         0.0          0.00                    0.0   

          total_amount  
17251945         35.34  
17251946         22.80  
17251947          0.00  
17251948          3.20  
17251949          0.00

In [7]:
# Checking Data from top
df60Days.head(5)

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2018-01-01 00:21:05   2018-01-01 00:24:23                1   
1         1  2018-01-01 00:44:55   2018-01-01 01:03:05                1   
2         1  2018-01-01 00:08:26   2018-01-01 00:14:21                2   
3         1  2018-01-01 00:20:22   2018-01-01 00:52:51                1   
4         1  2018-01-01 00:09:18   2018-01-01 00:27:06                2   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            0.5           1                  N            41            24   
1            2.7           1                  N           239           140   
2            0.8           1                  N           262           141   
3           10.2           1                  N           140           257   
4            2.5           1                  N           246           239   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             2          4.5    0.5      0.5        0.00           0.0   
1             2         14.0    0.5      0.5        0.00           0.0   
2             1          6.0    0.5      0.5        1.00           0.0   
3             2         33.5    0.5      0.5        0.00           0.0   
4             1         12.5    0.5      0.5        2.75           0.0   

   improvement_surcharge  total_amount  
0                    0.3          5.80  
1                    0.3         15.30  
2                    0.3          8.30  
3                    0.3         34.80  
4                    0.3         16.55

In [8]:
#Chekcing datatypes of variable before performing any analysis
df30Days.info() , df60Days.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8759874 entries, 0 to 8759873
Data columns (total 17 columns):
VendorID                 int64
tpep_pickup_datetime     object
tpep_dropoff_datetime    object
passenger_count          int64
trip_distance            float64
RatecodeID               int64
store_and_fwd_flag       object
PULocationID             int64
DOLocationID             int64
payment_type             int64
fare_amount              float64
extra                    float64
mta_tax                  float64
tip_amount               float64
tolls_amount             float64
improvement_surcharge    float64
total_amount             float64
dtypes: float64(8), int64(6), object(3)
memory usage: 1.1+ GB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17251950 entries, 0 to 17251949
Data columns (total 17 columns):
VendorID                 int64
tpep_pickup_datetime     object
tpep_dropoff_datetime    object
passenger_count          int64
trip_distance            float64
Ratec

(None, None)

#### tpep_pickup_datetime and tpep_dropoff_datetime are present as object insteat datetime datatype.So, timeConversion function is created.
    1. It converts the datatype to datetime.
    2. Fetch date,month,year,trip length for every ride and save into new columns.

In [9]:
#Converting object datatype of 'tpep_pickup_date' and 'tpep_dropoff_datetime' to datetime dataype
def timeConversion(df):
    if df.empty:
        print ("Please Pass DataFrame")
    else:
        df['tpep_pickup_datetime'] = pd.to_datetime(df.tpep_pickup_datetime)
        df['tpep_dropoff_datetime'] = pd.to_datetime(df.tpep_dropoff_datetime)
        df['year'] = df.tpep_pickup_datetime.dt.year
        df['month'] = df.tpep_pickup_datetime.dt.month
        df['day'] = df.tpep_pickup_datetime.dt.day
        df['tripLength'] = (df.tpep_dropoff_datetime -  df.tpep_pickup_datetime)
    return df

In [10]:
# Calling timeConversion function for datetimeconversion
df30 = timeConversion(df30Days)

In [11]:
# Calling timeConversion Function for datetime conversion
df60 = timeConversion(df60Days)

#### As mentioned  on the website of NYC TLC that data might not be accurate.So, function oneMonthCheck and twoMonthCheck are created to check the if we have data for year 2018 and months January, February respectively.

In [27]:
# Function to check if their is more than one year and month of data present in one month dataset
def oneMonthCheck(df):
    if df.empty:
        print ('Please Pass DataFrame')
    elif len(df.year.unique()) > 1 :
        print ('Please Check Years And Create a New DataFrame')
        print (df.year.unique())
    elif len(df.month.unique()) > 1 :
        print ('Please Check Months And Create a New DataFrame')
        print (df.month.unique())
    else:
        print ('DataFrame Is Good For Mean Calculations')
        print (df.year.unique())
        print ( df.month.unique())

In [13]:
# Calling the function to check total years and months present in dataset
oneMonthCheck(df30)

Please Check Years And Create a New DataFrame
[2018 2017 2002 2008 2009 2001 2003]


In [29]:
# Function to check if their is more than one year and 2 month of data present in 2 month dataset
def twoMonthCheck(df):
    if df.empty:
        print ('Please Pass DataFrame')
    elif len(df.year.unique()) > 1 :
        print ('Please Check Years And Create a New DataFrame')
        print (df.year.unique())
    elif len(df.month.unique()) > 2 :
        print ('Please Check Months And Create a New DataFrame')
        print (df.month.unique())
    else:
        print ('DataFrame Is Good For Mean Calculations')
        print (df.year.unique())
        print ( df.month.unique())

In [15]:
#Calling the function
twoMonthCheck(df60)

Please Check Years And Create a New DataFrame
[2018 2017 2002 2008 2009 2001 2003 2029 2019 2031 2026]


#### If we have data for more than one year and requested month for our analysis than use function oneMonthNew and twoMonthNew to create new dataframes as per our requirement.

In [16]:
# Function to prepare new dataset for one month as per requirements
def oneMonthNew(df,year,month):
    if df.empty:
        print ('Please Pass DataFrame')
    else:
        df = df[df.year== year][df.month == month]
    return df

In [17]:
#Calling the function
f30 = oneMonthNew(df30,2018,1)

In [28]:
#Calling the function to check if data is as per requirement
oneMonthCheck(f30)

DataFrame Is Good For Mean Calculations
[2018]
[1]


In [30]:
# Function to prepare new dataset for 45 days as per requirements
def twoMonthNew(df,year,month1,month2):
    if df.empty:
        print ('Please Pass DataFrame')
    else:
        df1 = df[df.year== 2018][df.month == month1]
        df2 = df[df.year== 2018][df.month == month2]
        df3 = df2[df2.day < 15]
        df4 = pd.concat([df1,df3],ignore_index= True)
    return df4

In [31]:
f45 = twoMonthNew(df60,2018,1,2)

In [32]:
twoMonthCheck(f45)

DataFrame Is Good For Mean Calculations
[2018]
[1 2]


#### Function tripMeanTrip is used to to calculate average time spent per trip.

In [33]:
def tripMeanTime(df):
    if df.empty:
        print ('Please Pass DataFrame')
    else:
        df.dropna()
    print(df.tripLength.mean())

In [34]:
tripMeanTime(f30)

0 days 00:15:39.371804


In [35]:
tripMeanTime(f45)

0 days 00:15:37.824905


#### Function tripMeanDistance is used to to calculate average  distance covered per trip.

In [36]:
def tripMeanDistance(df):
    if df.empty:
        print ('Please Pass DataFrame')
    else:
        df.dropna()
    print(df.trip_distance.mean())

In [37]:
tripMeanDistance(df30)

2.8040011340343773


In [38]:
tripMeanDistance(f45)

2.7678313902523657


## Results:
#### 1. Average time for 30 days is 00:15:39.371804 (hr:min:sec)
#### 2.  Average time for 45 days is 00:15:37.824905 (hr:min:sec)
#### 3.  Average distance covered for 30 days is 2.8039893295667246 miles
#### 4.  Average distance covered for 45 days is 2.7678313902523657 miles